### PR TITLE
satellite/metabase/rangedloop: minimal loop

### PR DESCRIPTION
**Pull request by Erikvv**
Minimal implementation of the ranged (=parallel) segment loop
service, to improve performance over the exiting loop.

Has tests with a an inmemory segment database
and example observer.

Does not have yet: database link, observer duration tracking,
suspicious processed ratio guard, rate limiting, minimum execution
interval per observer, etc.

Part of https://github.com/storj/storj/issues/5223

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/main/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/main/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/metabase/rangedloop/service_test.go
Comment (mniewrzal) [inline satellite/metabase/rangedloop/service_test.go:24]: I think we don't need `TestLoop_` prefix as this name will part of parent test. Now it will be `TestLoop/TestLoop_par1_seg0_obs0`.

File: satellite/metabase/rangedloop/service.go
Comment (mniewrzal) [inline satellite/metabase/rangedloop/service.go:91]: I think we don't need it here

File: satellite/metabase/rangedloop/service.go
Comment (mniewrzal) [inline satellite/metabase/rangedloop/service.go:109]: can observerStates[i] be just observerState? then `for _, observerState := range observerStates {`

File: satellite/metabase/rangedloop/service.go
Comment (mniewrzal) [inline satellite/metabase/rangedloop/service.go:132]: maybe errs or errList?

**Review by mniewrzal — APPROVED** [review]

File: satellite/metabase/rangedloop/service_test.go
Comment (Erikvv) [reply satellite/metabase/rangedloop/service_test.go:24]: Done

File: satellite/metabase/rangedloop/service.go
Comment (Erikvv) [reply satellite/metabase/rangedloop/service.go:109]: That doesn't work, because Go iterates by value, not by reference. `observerState` is a copy, so the change with `append()` is not visible outside the loop.

File: satellite/metabase/rangedloop/service.go
Comment (Erikvv) [reply satellite/metabase/rangedloop/service.go:132]: Done

**Comment by Erikvv** [comment]
race detector says there is a data race in `defer mon.Task()(&ctx)(&err)`, need to find out how to solve.

File: satellite/metabase/rangedloop/rangedlooptest/countobserver.go
Comment (mniewrzal) [inline satellite/metabase/rangedloop/rangedlooptest/countobserver.go:39]: ranged loop end?

**Review by mniewrzal — DISMISSED** [review]

File: satellite/metabase/rangedloop/rangedlooptest/countobserver.go
Comment (Erikvv) [reply satellite/metabase/rangedloop/rangedlooptest/countobserver.go:39]: Changed to "ranged segment loop"

That it processes segments is more important than that it works with ranges.

File: satellite/metabase/rangedloop/rangedlooptest/mockprovider.go
Comment (Fadila82) [inline satellite/metabase/rangedloop/rangedlooptest/mockprovider.go:14]: nit: I would call those ProviderMock and RangeProviderMock

File: satellite/metabase/rangedloop/service.go
Comment (Fadila82) [inline satellite/metabase/rangedloop/service.go:26]: Could we call it like in the other configs in metabase,ie. "AsOfSystemTime"?

File: satellite/metabase/rangedloop/service_test.go
Comment (Fadila82) [inline satellite/metabase/rangedloop/service_test.go:47]: typo: "parallelism"

File: scripts/testdata/satellite-config.yaml.lock
Comment (Fadila82) [inline scripts/testdata/satellite-config.yaml.lock:863]: typo: "parallelism"

File: satellite/metabase/rangedloop/service.go
Comment (Erikvv) [reply satellite/metabase/rangedloop/service.go:26]: If I grep for `AsOfSystem(.*)time.Duration` this is the only convention I can find in metabase.

File: satellite/metabase/rangedloop/rangedlooptest/mockprovider.go
Comment (Erikvv) [reply satellite/metabase/rangedloop/rangedlooptest/mockprovider.go:14]: Changed

**Review by mniewrzal — DISMISSED** [review]

**Review by Fadila82 — APPROVED** [review]

**Comment by Erikvv** [comment]
Renamed Provider interfaces to RangeSplitter and SegmentProvider after feedback in https://github.com/storj/storj/pull/5337. I would have had to do it during merge but I had to rebase this anyway.

Please re-approve.

**Review by mniewrzal — DISMISSED** [review]

**Comment by Erikvv** [comment]
TestDownloadSharesDownloadTimeout/Postgres failed, rerunning.

**Review by Fadila82 — DISMISSED** [review]

**Comment by azdagron** [comment]
This loop has some fairly different semantics than the segmentloop in terms of handling errors returned by observers. The segmentloop removes tends to remove observers that return failures, while this loop will end up failing out of the loop early. This means effectively that a runtime failure (or bug) in one observer can prevent other observers from ever making forward progress. For example, if the gracefulexit observer fails to flush to the transfer queue on a call to Join, then other observers who might have for all intents and purposes completed their observation loop successfully might never be Join'd or Finish'd.

Unless the semantics around Start/Fork/Join/Finish are that they should not return an error unless the intention is to abort the loop for _everybody_, we probably need some more robust handling of errors before turning this on, or at least redefine the error semantics around the Observer and Partial interface.

For example, I'd expect failures to Start and Fork to simply drop the observer from the _current_ cycle. I'd expect failures to Process to at _least_ stop calling process on the current partial, or maybe all partials for a given observer. It's unclear to me how failures to Join should be addressed. It's also hard for observers to know about failures for partials to process (which they may want to do something with in the Join or Finish call) unless Join is _always_ called for every partial even in the face of a failure to Process.

File: satellite/rangedloop.go
Comment (azdagron) [inline satellite/rangedloop.go:79]: Isn't this an import cycle?

Comment (Erikvv) [reply satellite/rangedloop.go:79]: It wouldn't compile if it was?

Comment (azdagron) [reply satellite/rangedloop.go:79]: I misread the filename at the top. I thought this was a file in the satellite/metabase/rangedloop package.

**Review by azdagron — DISMISSED** [review]

File: satellite/metabase/rangedloop/service.go
Comment (Erikvv) [reply satellite/metabase/rangedloop/service.go]: Done

**Review by azdagron — APPROVED** [review]

**Comment by Erikvv** [comment]
> This loop has some fairly different semantics than the segmentloop in terms of handling errors returned by observers. 

This was unintentional. I don't want to hold up this PR any longer. Created https://github.com/storj/storj/issues/5389

**Review by mniewrzal — APPROVED** [review]

**Comment by Erikvv** [comment]
TestAuditRepairedSegmentInExcludedCountries/Postgres failed seems unrelated. 

I want to rebase but I'm afraid I'll lose the approvals again. Asked Egon.

**Pull request merged**
